signature/jws/jws.go

```diff
@@ -0,0 +1,254 @@
+	err := validateProtectedHeaders(protectedHeader)
+	if err != nil {
+		return err
+	}
+
+	if signerInfo.SignatureAlgorithm, err = getSignatureAlgorithm(protectedHeader.Algorithm); err != nil {
+		return err
+	}
+
+	signerInfo.SignedAttributes.ExtendedAttributes = getExtendedAttributes(protectedHeader.ExtendedAttributes, protectedHeader.Critical)
+	signerInfo.SignedAttributes.SigningScheme = protectedHeader.SigningScheme
+	if protectedHeader.Expiry != nil {
+		signerInfo.SignedAttributes.Expiry = *protectedHeader.Expiry
+	}
+	switch protectedHeader.SigningScheme {
+	case signature.SigningSchemeX509:
+		if protectedHeader.SigningTime != nil {
+			signerInfo.SignedAttributes.SigningTime = *protectedHeader.SigningTime
+		}
+	case signature.SigningSchemeX509SigningAuthority:
+		if protectedHeader.AuthenticSigningTime != nil {
+			signerInfo.SignedAttributes.SigningTime = *protectedHeader.AuthenticSigningTime
+		}
+	}
+	return nil
+}
+
+func validateProtectedHeaders(protectedHeader *jwsProtectedHeader) error {
+	// validate headers that should not be present as per signing schemes
+	switch protectedHeader.SigningScheme {
+	case signature.SigningSchemeX509:
+		if protectedHeader.AuthenticSigningTime != nil {
+			return &signature.InvalidSignatureError{Msg: fmt.Sprintf("%q header must not be present for %s signing scheme", headerKeyAuthenticSigningTime, signature.SigningSchemeX509)}
+		}
+	case signature.SigningSchemeX509SigningAuthority:
+		if protectedHeader.SigningTime != nil {
+			return &signature.InvalidSignatureError{Msg: fmt.Sprintf("%q header must not be present for %s signing scheme", headerKeySigningTime, signature.SigningSchemeX509SigningAuthority)}
+		}
+		if protectedHeader.AuthenticSigningTime == nil {
+			return &signature.InvalidSignatureError{Msg: fmt.Sprintf("%q header must be present for %s signing scheme", headerKeyAuthenticSigningTime, signature.SigningSchemeX509)}
+		}
+	}
+
+	return validateCriticalHeaders(protectedHeader)
+}
+
+func validateCriticalHeaders(protectedHeader *jwsProtectedHeader) error {
+	if len(protectedHeader.Critical) == 0 {
+		return &signature.InvalidSignatureError{Msg: `missing "crit" header`}
+	}
+
+	mustMarkedCrit := map[string]bool{headerKeySigningScheme: true}
+	if protectedHeader.Expiry != nil && !protectedHeader.Expiry.IsZero() {
+		mustMarkedCrit[headerKeyExpiry] = true
+	}
+
+	if protectedHeader.SigningScheme == signature.SigningSchemeX509SigningAuthority {
+		mustMarkedCrit[headerKeyAuthenticSigningTime] = true
+	}
+
+	for _, val := range protectedHeader.Critical {
+		if _, ok := mustMarkedCrit[val]; ok {
+			delete(mustMarkedCrit, val)
+		} else {
+			if _, ok := protectedHeader.ExtendedAttributes[val]; !ok {
+				return &signature.InvalidSignatureError{
+					Msg: fmt.Sprintf("%q header is marked critical but not present", val)}
+			}
+		}
+	}
+
+	// validate all required critical headers are present.
```

Comment:
this comment can be more detailed, such as "// validate that all required headers(as per spec) are marked as critical"

Reply:
Updated.